### PR TITLE
refactoring document list

### PR DESCRIPTION
**Pull request by yneves**
* Created a reusable list component `TableList` and `Column` that leverage `React.cloneElement` to improve list definition.
 * Using `Immutable.Record` to define a `DocumentRecord` class to wrap documents.
 * Using `TableList` to generate documents list.

### CODE REVIEW

**Comment by yneves** [comment]
@ntgn81 This is an ongoing work, still needs lots of testing and fixes. Also, i will benchmark to evaluate rendering performance, but it looks promising. The `TableList` component will be good for the todo list as well. Let me know what you think.

**Comment by ntgn81** [comment]
Awesome! Can you walk me over this sometime later this week? On how the rendering performance can be better?

**Comment by yneves** [comment]
@ntgn81 Rendering performance hardly will get better, i'm happy if it doesn't get worse, that's why i'm measuring. The main point is making the code cleaner. Those lists have been bothering me since day one. But this change will unlock some further optimizations, since most props are passed directly from `DocumentsList` into the repeated element. For example, the list of suggestions in NPL descriptions, can be generated at the higher component and the same list reused by all rows. Being it all very transparent.

**Comment by yneves** [comment]
But yeah, would be nice to have a screen share and walk you through, for sure.

**Comment by yneves** [comment]
This is well tested now and also fixes a few recent bugs in document list. If you don't have any concerns, it's good to go. Unlike i was expecting, performance is actually better due to row change detection.

File: src/components/MultiSuggest.js [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import SimpleSuggest from './SimpleSuggest';
-import _get from 'lodash/object/get';
+import lodash from 'lodash';
 
 export default class MultiSuggest extends SimpleSuggest {
   getCursorPosition() {
-    const input = _get(this, 'refs.autoSuggest.refs.autoSuggest.input');
+    const input = lodash.get(this, 'refs.autoSuggest.refs.autoSuggest.input');
```
Comment (ntgn81) [inline src/components/MultiSuggest.js]: I used `_get` so the bundled file does not include everything from `lodash`, just `get`

File: src/components/documents/DocumentActions.js [hunk 2]
```diff
@@ -1,25 +1,38 @@
...
+  onAction(e, action) {
+    this.props.onAction(this.props.doc, action);
+  }
+
   render() {
+    if (!this.props.doc.isSaved()) {
+      return null;
+    }
+
     return (
```
Comment (ntgn81) [inline src/components/documents/DocumentActions.js]: Can we merge this into one? I think we have the same in other places:

```jsx
return this.props.doc.isSaved() && (
   <stuff>
   </stuff>
```

File: src/components/documents/DocumentCategories.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  onChange(e, selected) {
+    this.props.onChange(this.props.doc, this.props.field, selected);
+  }
+
+  render() {
+    if (!this.props.doc.isSaved()) {
+      return null;
+    }
+
+    return (
```
Comment (ntgn81) [inline src/components/documents/DocumentCategories.js]: Similar to comment above

File: src/components/documents/DocumentCheckbox.js [hunk 4]
```diff
@@ -34,20 +35,54 @@ const iconNormal = (
...
+
+  componentWillReceiveProps(nextProps) {
+    const checked = nextProps.doc.get(nextProps.field);
+    if (checked !== this.state.checked) {
+      this.setState({ checked });
+    }
+  }
+
+  onChange(e) {
+    const checked = e.target.checked;
```
Comment (ntgn81) [inline src/components/documents/DocumentCheckbox.js]: Does this have issues with being `undefined`? We had issues with that before

File: src/components/documents/DocumentCheckbox.js [hunk 5]
```diff
@@ -34,20 +35,54 @@ const iconNormal = (
...
+    this.setState({ checked });
+    this.props.onChange(this.props.doc, this.props.field, checked);
+  }
+
   render() {
+    if (!this.props.doc.isSaved()) {
+      return null;
+    }
+
     return (
```
Comment (ntgn81) [inline src/components/documents/DocumentCheckbox.js]: Same as above, can merge this in

**Comment by ntgn81** [comment]
Looking good, except for a few small comments! Go ahead and merge whenever you feel like it

**Pull request merged**